src/aws-cpp-sdk-core/include/aws/core/utils/stream/AwsChunkedStream.h

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#pragma once
+#include <aws/core/utils/HashingUtils.h>
+#include <aws/core/utils/Outcome.h>
+#include <aws/core/utils/crypto/Hash.h>
+
+namespace Aws {
+namespace Utils {
+namespace Stream {
+
+static const size_t AWS_DATA_BUFFER_SIZE = 65536;
+
+template <size_t DataBufferSize = AWS_DATA_BUFFER_SIZE>
+class AwsChunkedStream {
+ public:
+  AwsChunkedStream(Http::HttpRequest *request, const std::shared_ptr<Aws::IOStream> &stream)
+      : m_chunkingStream{Aws::MakeShared<StringStream>("AwsChunkedStream")}, m_request(request), m_stream(stream) {}
+
+  size_t BufferedRead(char *dst, size_t amountToRead) {
+    // the chunk has ended and cannot be read from
+    if (m_chunkEnd) {
+      return 0;
+    }
+
+    // If we've read all of the underlying stream write the checksum trailing header
+    // the set that the chunked stream is over.
+    if (m_stream->eof() && (m_chunkingStream->eof() || m_chunkingStream->peek() == EOF)) {
+      Aws::StringStream chunkedTrailer;
+      chunkedTrailer << "0\r\n";
+      if (m_request->GetRequestHash().second != nullptr) {
+        chunkedTrailer << "x-amz-checksum-" << m_request->GetRequestHash().first << ":"
+                       << HashingUtils::Base64Encode(m_request->GetRequestHash().second->GetHash().GetResult()) << "\r\n";
+      }
+      chunkedTrailer << "\r\n";
+      auto trailerSize = chunkedTrailer.str().size();
```

Comment:
perform str() extraction once
.str() is constructing a new std::string, so don't call .str() repeatedly
also, calling
```
chunkedTrailer.str().c_str()
```
will extract .c_str on a temp object, which lifetime will be extended, but still I'd prefer to be more explicit